### PR TITLE
Updatable owners

### PR DESCRIPTION
**Pull request by taco-paco**
Changes allow to update set of multisig owners and change number of required confirmations.

### CODE REVIEW

File: contracts/contracts/util.cairo [hunk 1]
```diff
@@ -0,0 +1,44 @@
+from starkware.cairo.common.math import assert_not_zero
+from starkware.cairo.common.default_dict import default_dict_new, default_dict_finalize
+from starkware.cairo.common.dict import dict_update
+from starkware.cairo.common.dict_access import DictAccess
+
+func almost_equal(a, b) -> (res : felt):
```
Comment (microbecode) [inline contracts/contracts/util.cairo:6]: Not used

File: contracts/hardhat.config.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ const config: HardhatUserConfig = {
   },
   networks: {
     devnet: {
-      url: "http://127.0.0.1:5050/",
+      url: 'http://127.0.0.1:5050/',
```
Comment (microbecode) [inline contracts/hardhat.config.ts:22]: All other places use double quotes, so we should stick to that. I added an issue to unify styling: https://github.com/eqlabs/starknet-multisig/issues/23

File: contracts/contracts/util.cairo [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    # default_value shall be not less than data_len
+    # otherwise data[default_value] can have duplicates for index in (default_value, data_len)
+    let (local my_dict_start) = default_dict_new(default_value=data_len)
+    let my_dict_end = my_dict_start
+
+    assert_unique_elements_impl{range_check_ptr=range_check_ptr, dict_ptr=my_dict_end}(
+        index=0, data_len=data_len, data=data
+    )
+
+    let (finalized_dict_start, finalized_dict_end) = default_dict_finalize(
```
Comment (microbecode) [inline contracts/contracts/util.cairo:39]: Hmm, I wonder if this is needed. The values are not used anyway - this function only needs to revert if there are duplicates, but otherwise I don't think it needs to be setting variables

File: contracts/contracts/Multisig.cairo [hunk 4]
```diff
@@ -322,80 +311,65 @@ end
...
-        owners_len : felt,
-        owners : felt*,
-        confirmations_required : felt,
-    ):
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*, confirmations_required : felt
+):
+    const lower_bound = 1
+
     with_attr error_message("invalid number of required confirmations"):
```
Comment (microbecode) [inline contracts/contracts/Multisig.cairo:319]: As far as I can tell, this is now duplicate code. This is checked in _set_confirmation_required_impl

File: contracts/contracts/util.cairo (hunk 3 above)
Comment (archseer) [reply contracts/contracts/util.cairo:39]: I pointed this out too but I wasn't sure if it's safe:

> Note that in order to enforce the consistency of subsequent dictionary accesses with the default values, one must eventually call default_dict_finalize() (which in turn calls dict_squash(), as discussed in the corresponding section). Otherwise, this is only enforced by hints which can be bypassed by a malicious prover.

File: contracts/contracts/Multisig.cairo [hunk 5]
```diff
@@ -470,24 +432,111 @@ func execute_transaction{
...
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    let (lt) = is_le(owners_len, confirmations_required - 1)
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
```
Comment (microbecode) [inline contracts/contracts/Multisig.cairo:506]: Shouldn't this, instead, revert? If someone tries to set 2 owners and 5 confirmations, the code simply silently changes 5 to 2. I would prefer the code not to do unexpected things silently. Setting 2 owners and 5 confirmation should be an error.

File: contracts/contracts/Multisig.cairo [hunk 6]
```diff
@@ -470,24 +432,111 @@ func execute_transaction{
...
+        ConfirmationsSet.emit(owners_len)
+
+        tempvar syscall_ptr = syscall_ptr
```
Comment (microbecode) [inline contracts/contracts/Multisig.cairo:509]: Oh how I love Cairo variable handling. Three times the same three lines, but I wouldn't be surprised if this is the only way to do it, because of...Cairo.

File: contracts/contracts/util.cairo [hunk 7]
```diff
@@ -0,0 +1,44 @@
+from starkware.cairo.common.math import assert_not_zero
```
Comment (microbecode) [inline contracts/contracts/util.cairo:1]: I think this file should have its own unit test file

File: contracts/test/Multisig.test.ts [hunk 8]
```diff
@@ -348,14 +348,9 @@ describe("Multisig with single owner", function () {
...
 
-      try {
-        await nonOwner.invoke(multisig, "execute_transaction", {
-          tx_index: txIndex,
-        });
-        expect.fail("Should have failed");
-      } catch (err: any) {
-        assertErrorMsg(err.message, "not owner");
-      }
+      await nonOwner.invoke(multisig, "execute_transaction", {
```
Comment (microbecode) [inline contracts/test/Multisig.test.ts:351]: shouldn't this revert?

File: contracts/contracts/util.cairo (hunk 3 above)
Comment (microbecode) [reply contracts/contracts/util.cairo:39]: Ah, interesting. Please at least add a link to the explanation next to the code.

**Comment by microbecode** [comment]
I'm ready with my first reviews, please go ahead and evaluate the feedback. Will possibly have another review round after the current feedback is resolved.

File: contracts/contracts/util.cairo (hunk 1 above)
Comment (taco-paco) [reply contracts/contracts/util.cairo:6]: Removed

File: contracts/contracts/util.cairo (hunk 3 above)
Comment (taco-paco) [reply contracts/contracts/util.cairo:39]: Removed unused return values. Added link and comment

File: contracts/contracts/util.cairo (hunk 1 above)
Comment (microbecode) [reply contracts/contracts/util.cairo:6]: Waiting for the commit. Thanks!

File: contracts/contracts/Multisig.cairo [hunk 9]
```diff
@@ -525,4 +571,27 @@ func _set_transaction_calldata{
...
+    _is_owner.write(owner_address, FALSE)
+    _owners.write(owners_index, 0)
+
+    return _clean_owners(owners_index + 1, owners_len)
+end
+
+func _set_confirmations_required_impl{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(confirmations_required : felt):
+    _confirmations_required.write(confirmations_required)
```
Comment (microbecode) [inline contracts/contracts/Multisig.cairo:593]: I feel the owner&confirmation number check should be here instead of in two separate places. Now it's both in constructor and in `set_confirmations_required`. And what makes it a bit more confusing is that they both now give a different error message.

File: contracts/contracts/Multisig.cairo (hunk 5 above)
Comment (microbecode) [reply contracts/contracts/Multisig.cairo:514]: Making this part of the review.

File: contracts/test/Multisig.test.ts (hunk 8 above)
Comment (microbecode) [reply contracts/test/Multisig.test.ts:351]: Making this part of the review.

File: contracts/test/Multisig.test.ts [hunk 10]
```diff
@@ -521,6 +516,195 @@ describe("Multisig with multiple owners", function () {
       assertErrorMsg(err.message, "need more confirmations");
     }
   });
+
+  // Tests below are interdependent and shall be run sequentially
+  it("transaction sets new owners", async function () {
```
[the remark this answers is not in the document]
Comment (microbecode) [reply contracts/test/Multisig.test.ts:521]: Making this part of the review.

**Review by microbecode — CHANGES_REQUESTED** [review]
Thanks for the changes! I have one extra comment, and included the remaining comments as part of a GitHub review. Sorry, I come from a different devops world, not yet used to GitHub style reviews.

File: contracts/test/Multisig.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (microbecode) [reply contracts/test/Multisig.test.ts:520]: In fact, now that I look at the unit tests structure, it's a mess and doesn't make much sense anyway. Just leave it as it is, I will restructure it all afterwards.

File: contracts/contracts/Multisig.cairo (hunk 4 above)
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo]: The same. Just realised that all my `pending` comments can not be seen.

Comment (microbecode) [reply contracts/contracts/Multisig.cairo]: Ok! Thanks for the explanation :)

File: contracts/test/Multisig.test.ts (hunk 8 above)
Comment (microbecode) [reply contracts/test/Multisig.test.ts:350]: This is an interesting approach. I assume you had an internal discussion about this - is it somewhere visible (in your own PRs for example)?
I see three options:
1) Keep it as it was before this PR
2) Change as you are suggesting
3) Make the transaction auto-execute upon last required confirmation.

If your idea is to make the process a bit easier (anyone can execute, so that could be performed by some untrusted backend of yours for example), then I would suggest option 3 instead. But do let me know about the reasoning behind this.

File: contracts/contracts/Multisig.cairo (hunk 5 above)
Comment (microbecode) [reply contracts/contracts/Multisig.cairo:514]: Ah you are correct, I didn't follow the logic about `set_owners_and_confirmations_required`. It will indeed revert if trying to set incompatible combination.
I don't have a very strong opinion about what should happen if amount of owner is lowered below required confirmation amount - should it revert or adjust the confirmation amount automatically. Can you share any internal reasoning you have behind this, or is it just based on coralxyz? In general I prefer to not have this kind of silent adjustments by the code and would maybe prefer a revert.

P.S. I'm very open to discussing these in a call or chat as well, if that makes things easier. Feel free to poke at Telegram @Lauri_P

File: contracts/test/Multisig.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/test/Multisig.test.ts:520]: Sure

File: contracts/contracts/util.cairo (hunk 7 above)
Comment (microbecode) [reply contracts/contracts/util.cairo]: Did you forget to commit "utils" file and the new test file?

File: contracts/contracts/Multisig.cairo (hunk 9 above)
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:593]: Agreed!

File: contracts/contracts/Multisig.cairo [hunk 11]
```diff
@@ -322,80 +311,65 @@ end
...
-    with_attr error_message("invalid number of required confirmations"):
-        assert_le(confirmations_required, owners_len)
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*, confirmations_required : felt
+):
+    const lower_bound = 1
+
+    with_attr error_message("range conditions unsatisfied"):
+        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
     end
```
Comment (krebernisak) [inline contracts/contracts/Multisig.cairo:321]: Let's move this to the internal `_set_confirmations_required` fn

File: contracts/contracts/Multisig.cairo [hunk 12]
```diff
@@ -470,24 +432,119 @@ func execute_transaction{
...
+):
+    alloc_locals
+    require_unique_owners(owners_len, owners)
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    let (lt) = is_le(owners_len, confirmations_required - 1)
```
[the remark this answers is not in the document]
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:511]: Also our condition should Ideally be `owners_len < confirmations_required` instead of `owners_len <= confirmations_required`

File: contracts/contracts/Multisig.cairo (hunk 5 above)
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:514]: Yeah, it's a simple workflow that was inspired by audited [Serum multisig](https://github.com/coral-xyz/multisig/blob/a413b76de3497273a0cd27bcba6c1ac439053b67/programs/multisig/src/lib.rs#L127) (now rebranded as Coral I guess).

The idea is:

- Owners and threshold can be set simultaneously (constructor, setter) or separately (setters).
- Owners can be added or removed. If removed so the total length is under the threshold, the threshold is updated to equal the new owners' length.
- Threshold can be increased or decreased, it should always be in the range `[1, owners_len]`

File: contracts/contracts/Multisig.cairo [hunk 13]
```diff
@@ -470,24 +432,119 @@ func execute_transaction{
...
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
+        ConfirmationsSet.emit(owners_len)
+
+        tempvar syscall_ptr = syscall_ptr
```
Comment (krebernisak) [inline contracts/contracts/Multisig.cairo:517]: Trying to understand this, why is this the only function where we update those builtins? Is it because the conditional case?

File: contracts/contracts/Multisig.cairo (hunk 5 above)
Comment (microbecode) [reply contracts/contracts/Multisig.cairo:514]: as I said, no strong opinion here. I'm ok with this

File: contracts/contracts/util.cairo (hunk 7 above)
Comment (taco-paco) [reply contracts/contracts/util.cairo]: Added them

File: contracts/contracts/Multisig.cairo [hunk 14]
```diff
@@ -322,80 +311,65 @@ end
...
+    const lower_bound = 1
+
+    with_attr error_message("range conditions unsatisfied"):
+        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
     end
 
-    _owners_len.write(value=owners_len)
-    _set_owners(owners_index=0, owners_len=owners_len, owners=owners)
-    _confirmations_required.write(value=confirmations_required)
+    _set_owners_impl(owners_len, owners)
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:323]: Well here I've added `_impl` suffix because `_set_owners` is already exists and used inside `_set_owners_impl`. I've mentioned that on slack that I could use some naming suggestions here. Don't like it either

File: contracts/contracts/Multisig.cairo [hunk 15]
```diff
@@ -470,24 +432,119 @@ func execute_transaction{
...
+end
+
+# Set new owners and number of required confirmations.
+# Can be called only via recursively from
+# execute_transaction -> set_owners_and_confirmations_required
+@external
+func set_owners_and_confirmations_required{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(owners_len : felt, owners : felt*, confirmations_required : felt):
+    set_owners(owners_len, owners)
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:476]: Sure this can be done via external call there is `call_contract` that can make recursive calls. But out curiosity: what is the reasoning behind making external call here? To avoid argument allocation in memory?

File: contracts/contracts/Multisig.cairo (hunk 13 above)
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:517]: This because of reference revocations. Cairo makes job of tracking variable shifts relative to current value of `ap` for us but it doesn't know which if-branch will be executed so it can't know how `ap` will advance in compile-time ergo it revokes reference.

https://www.cairo-lang.org/docs/how_cairo_works/consts.html

File: contracts/contracts/Multisig.cairo (hunk 12 above)
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:511]: > Also our condition should Ideally be owners_len < confirmations_required instead of owners_len <= confirmations_required

That's exactly what is done here: is_le checks if `a <= b` so `owners_len <= confirmations_required - 1` with integers is exactly `owners_len < confirmations_required`. Unfortunately there's no `is_lt` in cairo lib.

File: contracts/contracts/Multisig.cairo (hunk 14 above)
[the remark this answers is not in the document]
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:323]: Let's rename like this:

- `_set_owners_impl` -> `_set_owners`
- `_set_owners` -> `_set_owners_range`
- `_clean_owners` -> `_clean_owners_range`

File: contracts/contracts/Multisig.cairo (hunk 15 above)
[the remark this answers is not in the document]
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:476]: So my comment is suggesting the current workflow might produce external instead of internal (preferred, cheaper) calls. That's why I suggested functions should be refactored to make sure these calls are internal when calling `set_owners_and_confirmations_required` fn.

File: contracts/contracts/Multisig.cairo (hunk 12 above)
[the remark this answers is not in the document]
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:511]: Ok, that makes sense. Consider resolved!

But also consider adding a short inline comment explaining this logic, as it will make it easier to read/reason.

File: contracts/contracts/Multisig.cairo (hunk 9 above)
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:593]: I wouldn't agree with this. The point is to preserve invariants and to make sure any workflow makes the same checks, raises the same errors, and preserves all invariants (no surprises).

This is hard to do and follow if this kind of logic is spread all over the contract. Also introduces inconsistencies, like we currently have with error messages for this specific workflow.

Additionally, `_set_owners_impl` (similar fn) currently also checks owners are unique.

So in conclusion, it would be better if the `_set_confirmations_required` could include that range safety check (less code, stronger invariants, fewer bugs).

File: contracts/test/Multisig.test.ts (hunk 8 above)
Comment (krebernisak) [reply contracts/test/Multisig.test.ts:350]: The use case is for when one has a different set of signers (governance role) vs. executors (operations role). Yes, it also enables automation via protocols like Chainlink Keepers or custom infrastructure.

Option 2 vs. 3, we prefer option 2, as option 3 would require the last signer to pay gas fees for the transaction, which could be an order of magnitude larger than just voting on the proposal.

- It would introduce a game where signers would rush, then wait to avoid being last (probably a bad situation overall)
- It would make accounting more difficult, as there are use cases where operationally one doesn't want to pay for execution gas from the signer account but from another set of accounts.

This behavior is available from Serum multisig on Solana, and I also think Gnosis supports this workflow.

Comment (microbecode) [reply contracts/test/Multisig.test.ts:350]: Gnosis has this concept of "designated signer", who can propose transactions but not sign them - unsure of execution. But anyway, I now understand this better and it sounds fine. Thanks!

File: contracts/contracts/Multisig.cairo (hunk 9 above)
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:593]: > The point is to preserve invariants

Invariant is that values that are passed to setter is valid. Same is happening in `_set_owners` we've checked values before now we just set them.

> workflow makes the same checks, raises the same errors, and preserves all invariants (no surprises)

Its happening. Some inner function assume that values passed to them are correct, were checked and that's invariant. Business logic functions raise errors and so on.

> This is hard to do and follow if this kind of logic is spread all over the contract. 

Well my opinion is that if setter or getter does smth other than set or get that is break of Single Responsibility Principle. 

> Additionally, _set_owners_impl (similar fn) currently also checks owners are unique.

Similar function in that case would be `_set_owners`. `_set_owners_impl` does pre-checks and business logic and after that passes values to setter which as invariant assumes that values are valid

What I could suggest as compromise here is adding function smth like `verify_and_set_confirmations_required` which makes checks with `owners_len` and then calls `_set_confirmations_required`

Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:593]: I'm perfectly fine with adding the `verify_and_set_confirmations_required` 👍 

But, ideas like the Single Responsibility Principle and similar don't really map to blockchain contracts 1:1, because the environment is constrained. The focus is on security and cost (coupled with performance). You have limited and expensive resources, so best engineering practices have evolved to adapt to the environment (cost, tx limits, contract size limits, gas limits, etc.). It's similar to writing firmware code for constrained hardware devices. Sometimes there isn't enough space to be able to structure all the objects and functions as desired, so a tighter integration is required.

File: contracts/contracts/Multisig.cairo [hunk 16]
```diff
@@ -470,52 +426,131 @@ func execute_transaction{
...
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners_range(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    # checks if owners_len < confirmations_required
+    let (lt) = is_le(owners_len, confirmations_required - 1)
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
+        ConfirmationsSet.emit(owners_len)
```
Comment (krebernisak) [inline contracts/contracts/Multisig.cairo:504]: Let's use our setter here:

```suggestion
        _set_confirmations_required(owners_len)
```

File: contracts/contracts/Multisig.cairo [hunk 17]
```diff
@@ -470,52 +426,131 @@ func execute_transaction{
...
+    tempvar pedersen_ptr : HashBuiltin* = pedersen_ptr
+    tempvar range_check_ptr = range_check_ptr
+
+    let (tx_valid_since) = _next_tx_index.read()
+    _tx_valid_since.write(tx_valid_since)
+
+    _owners_len.write(owners_len)
+    # Recursively write the rest
+    _set_owners_range(0, owners_len, owners)
+    OwnersSet.emit(owners_len, owners)
```
Comment (krebernisak) [inline contracts/contracts/Multisig.cairo:524]: The order of operations in this function is strange:

1. Clean old owners
2. Set threshold
3. Set new owners

Let's consider changing it to something more expected and consistent with other workflows, so we preserve a consistent order of raised events (owners set first, then threshold).

1. Clean old owners
2. Set new owners
3. Set threshold

File: contracts/contracts/Multisig.cairo (hunk 16 above)
Comment (krebernisak) [inline contracts/contracts/Multisig.cairo:504]: Additionally, the way the code is currently structured it will produce 2x `ConfirmationsSet` events when calling `set_owners_and_confirmations_required` fn with fewer owners than the current threshold.

This might introduce bugs where a user might expect 2x events (owners + threshold set), but it will actually get 3x events.

Let's restructure code like this:

```cairo
@constructor
func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
    owners_len : felt, owners : felt*, confirmations_required : felt
):
    _set_owners_and_confirmations_required(owners_len, owners, confirmations_required)

    return ()
end 

@external
func set_owners_and_confirmations_required{
    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
}(owners_len : felt, owners : felt*, confirmations_required : felt):
    require_multisig()
    _set_owners_and_confirmations_required(owners_len, owners, confirmations_required) 
}

func _set_owners_and_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
    owners_len : felt, owners : felt*, confirmations_required : felt): 
    _set_owners(owners_len, owners)
    _validate_and_set_confirmations_required(confirmations_required, owners_len)
end

@external
func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
    owners_len : felt, owners : felt*
):
    require_multisig()
    _set_owners(owners_len, owners) // remove threshold check/set from here
    // TODO: check if new owners length less than current threshold, and if true call set_confirmations_required(owners_len)

    return ()
end
```

**Review by krebernisak — CHANGES_REQUESTED** [review]

File: contracts/contracts/Multisig.cairo (hunk 17 above)
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:524]: As discussed in the other comment, let's actually remove the threshold setting from here and move it (check + set) to the top-level `set_owners` function.

File: contracts/contracts/Multisig.cairo (hunk 15 above)
[the remark this answers is not in the document]
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:464]: We can resolve this comment and focus on the new code structure discussion that will also fix this issue: https://github.com/eqlabs/starknet-multisig/pull/22#discussion_r915872245

File: contracts/contracts/Multisig.cairo [hunk 18]
```diff
@@ -314,14 +314,8 @@ end
...
-    const lower_bound = 1
-
-    with_attr error_message("range conditions unsatisfied"):
-        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
-    end
-
-    _set_owners_impl(owners_len, owners)
-    _set_confirmations_required_impl(confirmations_required)
+    _validate_and_set_confirmations_required(confirmations_required, owners_len)
+    _set_owners(owners_len, owners)
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:318]: `_validate_and_set_confirmations_required` also validates `owners_len` so it makes sense to call it first

File: contracts/contracts/Multisig.cairo (hunk 17 above)
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:515]: Consider resolved!

File: contracts/contracts/Multisig.cairo (hunk 16 above)
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo]: Consider resolved!

File: contracts/contracts/Multisig.cairo (hunk 18 above)
[the remark this answers is not in the document]
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:318]: Please switch the order of these calls so they are consistent throughout the contract.

1. Constructor should set the owners first, then threshold (currently opposite)
2. `set_owners` fn should set the owners first, update the threshold if new owners are less than the existing threshold (currently ok)
3. `set_owners_and_threshold` fn should set owners first, update the threshold second (currently opposite)

File: contracts/contracts/Multisig.cairo (hunk 16 above)
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo]: Consider resolved (partly for now, but we can resolve this comment)!

**Review by krebernisak — APPROVED** [review]

File: contracts/contracts/Multisig.cairo (hunk 18 above)
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:318]: > Constructor should set the owners first, then threshold (currently opposite)

Doesn't it make sense to first check that `owners_len` and `confirmations_required` satisfy following requirements:
`owners_len` not zero
`confirmations_required` is within [1, owners_len]. Because otherwise there is no point to proceed with setting threshold or owners
And than call set_owners which checks uniqueness owners and so on?

> set_owners_and_threshold fn should set owners first, update the threshold second (currently opposite)

Same as above

[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:318]: Well [if this ](https://github.com/eqlabs/starknet-multisig/pull/22#discussion_r916683876) is not a case then swapping places is ok

[the remark this answers is not in the document]
Comment (krebernisak) [reply contracts/contracts/Multisig.cairo:318]: It makes sense to save some gas like you are suggesting.

> Following order can be achieved only by splitting _validate_and_set_confirmations_required on 2 functions: validate & set

IMHO this is the way to go.

1. Remove the `_validate_and_set_confirmations_required` fn altogether
2. Replace it with a standard require fn, for example `require_confirmations_required_valid` + existing setter
3. In constructor and `set_owners_and_threshold` functions validate the requirement first, then set owners and threshold

```cairo
require_confirmations_required_valid(confirmations_required, owners_len)
_set_owners(owners_len, owners)
_set_confirmations_required(confirmations_required)
```

[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo:318]: Great! Will do

[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/contracts/Multisig.cairo]: Done

**Review by krebernisak — APPROVED** [review]

**Review by microbecode — APPROVED** [review]
Looks good! Thanks everyone for the hard work.

**Comment by microbecode** [comment]
Merging
